nsqd/channel.go

```diff
@@ -196,16 +231,25 @@ func (c *Channel) Empty() error {
 		client.Empty()
 	}
 
+MemoryDrain:
 	for {
 		select {
 		case <-c.memoryMsgChan:
 		default:
-			goto finish
+			break MemoryDrain
 		}
 	}
 
-finish:
-	return c.backend.Empty()
+	err := c.backend.Empty()
+	if err != nil {
+		return err
```

Comment:
good catch.  The go-diskqueue will always set the in-process state to zero even if it's unable to remove files (i.e. if the disk is readonly); So i guess the right thing to do is always do the `c.isDraining` check, otherwise you would have to re-empty (even though it appeared there were no messages) to finish the drain.

https://github.com/nsqio/go-diskqueue/blob/5d0854e78c3360a46dfb324d6f09fac0bb2507d9/diskqueue.go#L267